presto-tests/src/test/java/io/prestosql/tests/TestMinWorkerRequirement.java

```diff
@@ -69,18 +107,136 @@ public void testInsufficientWorkerNodesAfterDrop()
+            throws Exception
+    {
+        try (DistributedQueryRunner queryRunner = TpchQueryRunnerBuilder.builder()
+                .setCoordinatorProperties(ImmutableMap.<String, String>builder()
+                        .put("query-manager.required-workers", "3")
+                        .put("query-manager.required-workers-max-wait", "1ns")
+                        .build())
+                .setNodeCount(2)
+                .build()) {
+            Session session = testSessionBuilder()
+                    .setSystemProperty(REQUIRED_WORKERS_COUNT, "3")
+                    .setSystemProperty(REQUIRED_WORKERS_MAX_WAIT_TIME, "99ns")
+                    .setCatalog("tpch")
+                    .setSchema("tiny")
+                    .build();
+            queryRunner.execute(session, "SELECT COUNT(*) from lineitem");
+            fail("Expected exception due to insufficient active worker nodes");
+        }
+    }
+
+    @Test
+    public void testRequiredWorkerNodesSessionOverride()
+            throws Exception
+    {
+        try (DistributedQueryRunner queryRunner = TpchQueryRunnerBuilder.builder()
+                .setCoordinatorProperties(ImmutableMap.<String, String>builder()
+                        .put("query-manager.required-workers", "5")
+                        .put("query-manager.required-workers-max-wait", "1ns")
+                        .build())
+                .setNodeCount(4)
+                .build()) {
+            // Query should be allowed to run if session override allows it
+            Session session = testSessionBuilder()
+                    .setSystemProperty(REQUIRED_WORKERS_COUNT, "4")
+                    .setCatalog("tpch")
+                    .setSchema("tiny")
+                    .build();
+            queryRunner.execute(session, "SELECT COUNT(*) from lineitem");
+
+            // Query should not be allowed to run because we are 2 nodes short of requirement
+            session = Session.builder(session)
+                    .setSystemProperty(REQUIRED_WORKERS_COUNT, "6")
+                    .build();
+            try {
+                queryRunner.execute(session, "SELECT COUNT(*) from lineitem");
+                fail("Expected exception due to insufficient active worker nodes");
+            }
+            catch (RuntimeException e) {
+                assertEquals(e.getMessage(), "Insufficient active worker nodes. Waited 1.00ns for at least 6 workers, but only 4 workers are active");
+            }
+
+            // After adding 2 nodes, query should run
+            queryRunner.addServers(2);
+            assertEquals(queryRunner.getCoordinator().refreshNodes().getActiveNodes().size(), 6);
+            queryRunner.execute(session, "SELECT COUNT(*) from lineitem");
+        }
+    }
+
+    @Test
+    public void testMultipleRequiredWorkerNodesSessionOverride()
+            throws Exception
+    {
+        ExecutorService queryExecutor = newFixedThreadPool(3);
```

Comment:
There is a much simpler way to do the background work:
```
ListeningExecutorService queryExecutor = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(1));

ListenableFuture<ResultWithQueryId<MaterializedResult>> queryFuture1 = service.submit(() -> queryRunner.executeWithQueryId(session,  "SELECT COUNT(*) from lineitem"));
```
With that code, we no longer need the new `executeAsync` on `DistributedQueryRunner`.

BTW. The Guava site has excellent documentation on `ListenableFuture` usage: https://github.com/google/guava/wiki/ListenableFutureExplained#creation

Reply:
Thanks, I've now simplified this to use `ListeningExecutorService` and removed `executeAsync` from `DistributedQueryRunner`